src/Microsoft.ML.Api/SchemaDefinition.cs

```diff
@@ -312,8 +312,9 @@ public void RemoveMetadata(string kind)
         /// Create a schema definition by enumerating all public fields of the given type.
         /// </summary>
         /// <param name="userType">The type to base the schema on.</param>
+        /// <param name="vectorSizes"></param>
         /// <returns>The generated schema definition.</returns>
-        public static SchemaDefinition Create(Type userType)
+        public static SchemaDefinition Create(Type userType, Dictionary<string, int[]> vectorSizes = null)
```

Comment:
>public static SchemaDefinition Create(Type userType, Dictionary<string, int[]> vectorSizes = null) [](start = 8, length = 98)

Is it possible you've misunderstood the idiom? This is not how this method was intended to be used. The reason why the `SchemaDefinition` object is mutable is *precisely* to allow for this scenario. The idea is, if you want to infer a schema definition over a type, you're free to do so, and then in a separate process you can assign the vector lengths to the types.

In the "new" entry-point API we've actually regressed considerably, by making inference-from-type a one step feature. That misdesign is the source of the problem, not the fact that we didn't think to pass a `Dictonary<string, int[]>` everywhere. :) Besides there are more to runtime types than merely what the vector size is.

Reply:
I suggest we not complicate this code any more than we need to. This code, we know from experience, already solves this problem successfully. I'd confine any "adjustments" necessary to the code in the existing API, though I do not think the time apt: this will not go in for v0.3 I am fairly certain, which means that enhancements are for a doomed scenario anyway.

---
In reply to: [199036002](https://github.com/dotnet/machinelearning/pull/449#discussion_r199036002) [](ancestors = 199036002)

Reply:
I knew I was doing something wrong, but this days it's easier to write code and send it for review, rather than find person and ask...
I agree this is a dead code, but in same time, I would say it's a bit hard to figure out all this schema, it's structure, and first of all to find this method.
Hope we will put examples of how users can create SchemaDefinition and set vector sizes in our examples :)


---
In reply to: [199036302](https://github.com/dotnet/machinelearning/pull/449#discussion_r199036302) [](ancestors = 199036302,199036002)